src/cdk/testing/harness-environment.ts

```diff
@@ -147,43 +130,179 @@ export abstract class HarnessEnvironment<E> implements HarnessLoader, LocatorFac
+    // Combine all of the queries into one large comma-delimited selector and use it to get all raw
+    // elements matching any of the individual queries.
+    const rawElements = await this.getAllRawElements(
+        [...elementQueries, ...harnessQueries.map(pred => pred.getSelector())].join(','));
+    // If there are only element queries, we can just return a `TestElement` for each raw element
+    // that was matched.
+    if (harnessQueries.length === 0) {
+      return this._matchRawElementsToTestElements(rawElements) as LocatorFnResult<T>[];
+    }
+    // If the queries contain only searches for a single harness type, we can just create instances
+    // of that class for all raw elements and filter out the ones that don't match any query.
+    if (elementQueries.length === 0 && harnessTypes.size === 1) {
+      return this._matchRawElementsToHarnessType(
+          rawElements, harnessQueries[0].harnessType, harnessQueries);
+    }
+    // Otherwise the queries are come combination of different ComponentHarness classes and string
+    // selectors. In this case we need to do extra matching ahead of time to determine which
+    // selectors were responsible for finding which raw elements.
+    return this._matchRawElementsToHarnessesAndTestElements(rawElements, allQueries);
+  }
+
+  /** Matches the given raw elements with test elements. */
+  private _matchRawElementsToTestElements(rawElements: E[]): TestElement[] {
+    return rawElements.map(rawElement => this.createTestElement(rawElement));
+  }
+
+  /**
+   * Matches the given raw elements with a list of harness predicates that all produce the same type
+   * of harness to produce a list of matched harnesses.
+   */
+  private async _matchRawElementsToHarnessType<T extends ComponentHarness>(
+      rawElements: E[], harnessType: ComponentHarnessConstructor<T>,
+      harnessPredicates: HarnessPredicate<T>[]): Promise<T[]> {
+    return (
+        await Promise.all(rawElements.map(rawElement => {
+          const harness = this.createComponentHarness(harnessType, rawElement);
+          // Get a list of boolean results by comparing the harness against each predicate, then
+          // check if the harness matched any of them.
+          return Promise.all(harnessPredicates.map(pred => pred.evaluate(harness)))
+              .then(matches => matches.some(isMatch => isMatch) ? harness : null);
+        }))
+    ).filter((harness): harness is T => !!harness);
+  }
+
+  /**
+   * Matches the given raw elements with the given list of element and harness queries to produce a
+   * list of matched harnesses and test elements.
+   */
+  private async _matchRawElementsToHarnessesAndTestElements<T extends
+      (string | HarnessPredicate<any>)[]>(rawElements: E[], allQueries: T):
+      Promise<(LocatorFnResult<T>)[]> {
+    return Promise.all(rawElements.map(rawElement => {
+      const testElement = this.createTestElement(rawElement);
+      return Promise.all(
```

Comment:
You could use async/await anywhere where you have `.then`.. E.g.

```
const perElementMatches = await Promise.all(rawElements.map(async rawElement => {
  const testElement = this.createTestElement(rawElement);
  const allResultsForElement = await Promise.all(...);
  return _removeDuplicateQueryResults(allResultsForElement));
}));

return [...perElementMatches];
```

Reply:
I think started avoiding that because `filter` didn't work with `async`/`await` but since I wound up just mapping everything I think you're right. Will update...